pkg/debuginfo/debuginfo.go

```diff
@@ -53,12 +57,49 @@ func New(logger log.Logger, client Client, tmp string) *DebugInfo {
 		),
 		debugInfoFileCache: cache.New(cache.WithMaximumSize(128)), // Arbitrary cache size.
 		client:             client,
-		Extractor:          NewExtractor(logger, client, tmp),
-		Uploader:           NewUploader(logger, client),
-		Finder:             NewFinder(logger),
+		// Up to this amount of debug files in flight at once. This number is very large
+		// and unlikely to happen in real life.
+		uploadingCache: cache.New(
+			cache.WithMaximumSize(1024),
+		),
+		Extractor: NewExtractor(logger, client, tmp),
+		Uploader:  NewUploader(logger, client),
+		Finder:    NewFinder(logger),
 	}
 }
 
+// We upload the debug information files concurrently. In case
+// of two files with the same buildID are extracted at the same
+// time, they will be written to the same file.
+//
+// Most of the time, the file is, erm, eventually consistent-ish,
+// and once all the writers are done, the debug file looks is an ELF
+// with the correct bytes.
+//
+// However I don't believe there's any guarantees on this, so the
+// files aren't getting corrupted most of the time by sheer luck.
+//
+// These two helpers make sure that we don't try to extract + upload
+// the same buildID concurrently.
+func (di *DebugInfo) alreadyUploading(buildID string) bool {
+	di.uploadingCacheMutex.Lock()
```

Comment:
Do we really need these locking mechanisms? The `cache.Cache` looks to me as if it's safe for concurrent use.

Reply:
Good catch! https://github.com/goburrow/cache/issues/22#issuecomment-791784927